File: leetcode/0938-range-sum-of-bst/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int sum, low, high;
    public int rangeSumBST(TreeNode root, int low, int high) {
        sum = 0;
        this.low = low;
        this.high = high;
        
        traverse(root);
        
        return sum;
    }
    
    private void traverse(TreeNode root) {
        if (root == null)
            return;
        
        if (low <= root.val && root.val <= high) {
            sum += root.val;
            traverse(root.left);
            traverse(root.right);
        }
        
        if (root.val < low)
            traverse(root.right);
        
        if (high < root.val)
            traverse(root.left);
    }
}

File: leetcode/0938-range-sum-of-bst/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        if (root == NULL)
            return 0;
        
        if (low <= root->val && root->val <= high)
            return root->val + rangeSumBST(root->left, low, high) + rangeSumBST(root->right, low, high);
        
        if (root->val < low)
            return rangeSumBST(root->right, low, high);
        
        return rangeSumBST(root->left, low, high);
    }
};